src/palimpzest/query/operators/aggregate.py

```diff
@@ -664,3 +689,245 @@ def __call__(self, candidates: list[DataRecord]) -> DataRecordSet:
         )
 
         return DataRecordSet([dr], [record_op_stats])
+    
+class SemanticGroupByOp(AggregateOp):
+    """
+    Implementation of a semantic GroupBy operator using LLMs. This operator groups records by a set 
+    of fields and applies aggregation functions to each group using an LLM to determine the groups.
+    """
+    def __init__(self, gby_fields: list[str], agg_fields: list[str], agg_funcs: list[str], 
+                 model: Model | None = None, prompt_strategy: PromptStrategy = PromptStrategy.AGG, 
+                 reasoning_effort: str | None = None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.gby_fields = gby_fields
+        self.agg_fields = agg_fields
+        self.agg_funcs = agg_funcs
+        self.model = model if model is not None else Model.GPT_4O_MINI
+        self.prompt_strategy = prompt_strategy
+        self.reasoning_effort = reasoning_effort
+        
+        # Initialize the generator for LLM calls
+        self.generator = Generator(self.model, self.prompt_strategy, self.reasoning_effort, self.api_base)
+
+    def __str__(self):
+        op = super().__str__()
+        op += f"    Group-by Fields: {self.gby_fields}\n"
+        op += f"    Agg. Fields: {self.agg_fields}\n"
+        op += f"    Agg. Funcs: {self.agg_funcs}\n"
+        op += f"    Model: {self.model.value}\n"
+        op += f"    Prompt Strategy: {self.prompt_strategy}\n"
+        return op
+
+    def get_id_params(self):
+        id_params = super().get_id_params()
+        return {
+            "gby_fields": self.gby_fields, 
+            "agg_fields": self.agg_fields, 
+            "agg_funcs": self.agg_funcs,
+            "model": self.model.value,
+            "prompt_strategy": self.prompt_strategy.value,
+            "reasoning_effort": self.reasoning_effort,
+            **id_params
+        }
+
+    def get_op_params(self):
+        op_params = super().get_op_params()
+        return {
+            "gby_fields": self.gby_fields, 
+            "agg_fields": self.agg_fields, 
+            "agg_funcs": self.agg_funcs,
+            "model": self.model,
+            "prompt_strategy": self.prompt_strategy,
+            "reasoning_effort": self.reasoning_effort,
+            **op_params
+        }
+    
+    def get_model_name(self) -> str:
+        return self.model.value
+    
+    def get_fields_to_generate(self, candidate: DataRecord) -> list[str]:
+        """
+        For aggregation operators, we need to generate ALL output fields (including group-by fields),
+        not just the new fields. This overrides the default behavior.
+        """
+        return list(self.output_schema.model_fields.keys())
```

Comment:
One quirk about PZ's schemas is that they are additive. For example, if I do:
```
ds = pz.Dataset(...)
ds1 = ds.sem_map(cols=[{"name": "summary", ...}])
ds2 = ds1.sem_map(cols=[{"name": "title", ...}])
```
The `self.output_schema.model_fields` for the second `sem_map` would contain both `summary` and `title`; however this operation only generates `title`!

Taken together with my comment above re: keeping the `agg_field_name = agg_fields[i]` the base class implementation of `PhysicalOperator.get_fields_to_generate()` _should (in theory)_ suffice here. Thus, I think you can safely delete this function.